src/messenger/messenger.cpp

```diff
@@ -398,128 +404,225 @@ void Messenger::Terminate() {
   is_messenger_running_ = false;
 }
 
-void Messenger::ListenForConnection(const ConnectionDestination &target, const std::string &identity,
-                                    CallbackFn callback) {
+router_id_t Messenger::ListenForConnection(const ConnectionDestination &target, const std::string &identity,
+                                           CallbackFn callback) {
   // ZeroMQ is not thread-safe, and the actual binding of new connection endpoints must be done from the same
   // thread that is going to poll the endpoints. See RouterToBeAdded docstring.
   std::unique_lock lock(routers_add_mutex_);
-  routers_to_be_added_.emplace_back(RouterToBeAdded{target, identity, std::move(callback)});
+  router_id_t router_id = next_router_id_++;
```

Comment:
I'm not sure if this matters, but since `next_router_id_` is only accessed with a lock it might not need to be an atomic int.

Reply:
same as above, changed